ENH: simplify field indexing of structured arrays

ahaldane opened the pull request:
This commit simplifies the code in `array_subscript` and `array_assign_subscript` related to field access. This fixes https://github.com/numpy/numpy/issues/4806, and also removes a potential segfaults, eg if the array is indexed using an sequence-like object that raises an exception in getitem.
#### KeyError vs ValueError

One minor comment on field subscripts: Currently the following code raises a `ValueError`:

```
>>> a = np.zeros((1,), dtype=[('f1', 'i4')]
>>> a['nofield']
```

However `KeyError` seems slightly better to me, and it's what my code "naturally" produced (but I corrected it since some tests expected `ValueError`). Is it worth changing? (Probably not, for back-compatability).

(Also note that multi-field indexing still returns a copy, though it had been planned to return a view in 1.10. I originally considered changing that here, but I left that for another time).

ahaldane committed:
ENH: simplify field indexing of structured arrays

This commit simplifies the code in array_subscript and
array_assign_subscript related to field access. This fixes #4806,
and also removes a potential segfaults, eg if the array is indexed using
an sequence-like object that raises an exception in getitem.

Also fixes #5631, related to creation of structured dtypes
with no fields (an unusual and probably useless edge case).

Also moves all imports in _internal.py to the top.

Fixes #4806.
Fixes #5631.

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1735,36 +1725,30 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
-            PyObject *obj;
+    /* field access */
+    if (PyDataType_HASFIELDS(PyArray_DESCR(self)) &&
+            obj_is_string_or_stringlist(ind)) {
```
Just this one note for the moment. This seems wrong, we do not support assignmnet to multiple fields at once. You could do this (but think it might be a bad idea, considering how bad multi field assignments are right now), if you give the python function an option to return a view already now. Otherwise we are basically assigning into dev NULL ;).

ahaldane replied:
Ah, I had intended for the following two examples to behave similarly.

    >>> a = np.array([(1,2,3),(4,5,6)], dtype=[('f0', 'i8'), ('f1', 'i8'), ('f2', 'i8')])
    >>> b = a[['f0', 'f2']]
    >>> b[:] = (10,20)

    >>> a = np.array([(1,2,3),(4,5,6)], dtype=[('f0', 'i8'), ('f1', 'i8'), ('f2', 'i8')])
    >>> a[['f0', 'f2']] = (10,20)

The first example was already possible - multi-field indexing returns a copy, and shows a warning if you try to write to it. With this PR, I wanted second example to also show a warning. (Both examples effectively write to dev/null).

I forgot to show the warning though. I will add it in `array_assign_subscript`.

I don't think multi-field indexing should return a view (even as an option) yet. I tried doing so, but indeed, as you mention, I discovered that it would behave quite strangely. I think more extensive changes are needed before multi-field assignment is really allowed.

ahaldane commented:
All right, I added the warning. It is easy to switch back to forbidding multi-field subscript assignment if that's better. The advantage with the current code is that once multi-field indexing becomes feasible all that is needed is to remove the warning. A possible downside is the warning is shown on every assignment attempt (I didn't see a way to prevent this).

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1735,35 +1726,40 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
+        /* warn if writing to a copy. copies will have no base */
+        if (PyArray_BASE(view) == NULL) {
+            /* the warning will be shown on every attempt */
+            PyArray_ENABLEFLAGS((PyArrayObject*)view, NPY_ARRAY_WARN_ON_WRITE);
```
This sounds like it just warns, it needs to raise an error. I do not think I would care about the future warning, if we want to change it, we can just change it. Your example will work fine also for the second case, that is not the thing, the first case is just also broken enough in my opinion, but then I don't like recarrays.
In any case, I am not sure the changes to this part give any gain at all to be honest, unless you want the multi index thing.

charris commented:
@ahaldane I did have a branch for taking views from structure, but a test failed due to changes in filler. I thought I made a post, but can't find it now. In any case, the result of the change was not completely back compatible, but the failures would probably be few.

ahaldane commented:
@seberg, I did have multi-field assignment in mind, but since that isn't implemented it's true the changes to `array_assign_subscript` don't do much.

I'm fine with changing `array_assign_subscript` back, but let me try one other variation first: I now raise an error on multi-field assignment attempt. The difference is the error message now says "multi-field assignment is not supported" but it used to say "only integers, slices (`:`), ellipsis (`...`), numpy.newaxis (`None`) and integer or boolean arrays are valid indices", and raises ValueError instead of IndexError.

ahaldane commented:
@charris Alignment issues hadn't occurred to me. Thanks, I'll keep it in mind.

seberg commented:
Throwing a new custom IndexError seems like a great change to me. Calling the python code is fine with me, the only thing I could disagree with would be speed reasons, and to be honest I doubt they are large, and I don't care about recarray speed ;).
.

ahaldane commented:
Sounds good!

I added another small change as a second commit, to fix #5631.

charris commented:
@seberg Are you suggesting an `IndexError` be added to `numpy/__init__.py`? If so it should also be mentioned in the release notes.

seberg commented:
Oh sorry, "custom" wasn't meant like that, no. I just wrote that because before it dropped through to the normal indexing for which fields don't make sense really. Should maybe mention a chane in error type anyway.

charris commented:
@ahaldane Needs rebase. @seberg comments?

ahaldane commented:
rebased + tidied up.

Here's the current summary:

1) Tweaked `_index_fields`, so that now it handles individual field indices, and also does more careful sanity checks (fixing #4806). It is a small step closer to allowing multi-field assignment, if that will ever happen.

2) Simplified `array_subscript` and `array_assign_subscript`: It now leaves most of the work to `_index_fields`, and raises a more descriptive error on attempts to do multi-field assignment. The check for field indexing is now a separate function `obj_is_string_or_stringlist`, which fixes a potential segfault present before. I also just updated it to use the new `npy_cache_pyfunc`.

3) Bugfix in `VOID_setitem`, to fix #5631

seberg commented on numpy/core/_internal.py:
```diff
@@ -287,23 +287,45 @@ def _newnames(datatype, order):
         return tuple(list(order) + nameslist)
     raise ValueError("unsupported order value: %s" % (order,))
 
+def isstring(obj):
```
I would prefer a conditional def instead of a condition inside the function. Do we actually support unicode fields on python 2? If not, I think you can just use `str`, if we do, you can use `six.string_types` instead.

seberg commented on numpy/core/_internal.py:
```diff
@@ -287,23 +287,45 @@ def _newnames(datatype, order):
-    view = ary.view(dtype=view_dtype)
+    for name in names:
+        if name not in dt.fields:
+            raise ValueError("field name %s not found" % name)
```
I think this is an IndexError probably?

seberg replied:
Oh, right it would be KeyError. I think we can probably change it to be honest. I changed a couple of weird value to index error last version and nobody seemed to have noticed.

seberg commented on numpy/core/_internal.py:
```diff
@@ -287,23 +287,45 @@ def _newnames(datatype, order):
+        try:
+            return ary.getfield(dt.fields[names][0], dt.fields[names][1])
+        except KeyError as e:
+            # outside code expects valueerror in field lookup failure
```
Or maybe not index error, but which outside code?!

seberg commented:
I will have a better look tomorrow morning probably. Don't have time right now, and more annoying stuff happened also ;).

ahaldane commented:
Modified `isstring` as suggested. And while in python2 fields cannot be unicode, field titles can be, so I do have to take care of unicode.

I also added a provisional commit changing the `ValueError` to a `KeyError` for invalid field access. It required tweaking 4 unit tests plus a line in `lib/recfunctions.py`. I would definitely prefer it to be a `KeyError`, but a github search did turn up one case expecting `ValueError` , (here)[https://github.com/sjoertvv/sjoert/blob/f849373f23ae3e19db0caab1f5cd693d82600d55/rec.py#L92]. Any opinions?

seberg commented on numpy/core/_internal.py:
```diff
@@ -287,23 +287,38 @@ def _newnames(datatype, order):
         return tuple(list(order) + nameslist)
     raise ValueError("unsupported order value: %s" % (order,))
 
+if sys.version_info >= (3, 0):
```
Sorry, the real version I liked was to just use six.. Which is actualyl incorrect, it is to use:
```
from numpy.compat import basestring
```
and then just acting as if you are in python 2, so you can remove this completly.

seberg commented:
About the key error. I would prefer if we can make the change somehow. Maybe a mail to the list can gather some vetos ;). It would definitely require a note in the Release Notes. Not sure if one broken place warrents the change (and it is at least a gracious fail), but then maybe there is more and some larger package breaks....

ahaldane commented:
Oh, very nice. Good to know about `numpy.compat`. 

I'll write the list on the `KeyError` issue.

seberg commented on numpy/core/_internal.py:
```diff
@@ -288,22 +288,32 @@ def _newnames(datatype, order):
+    view_dtype = {'names': names, 'formats': formats,
+                  'offsets': offsets, 'itemsize': dt.itemsize}
+
+    #return copy for now (future plan to return ary.view(dtype=view_dtype))
```
Space after # ;), next line also misses some spaces, but it isn't your line

seberg commented on numpy/core/_internal.py:
```diff
@@ -288,22 +288,32 @@ def _newnames(datatype, order):
-    from .multiarray import empty, dtype, array
+# construct new array with just those fields
+def _index_fields(ary, names):
+    from .multiarray import dtype, array
```
Just noticed, will merge without, but if you are bored and want to increase your commit count.... We should probably move all these imports here to the module level.

ahaldane replied:
Hmm I could have sworn I added this to fix a bug. Maybe it was in an old version of this PR where `_index_fields` took care of more cases. In any case I'll remove it now.

seberg commented on numpy/core/_internal.py:
```diff
@@ -288,22 +288,32 @@ def _newnames(datatype, order):
-    names = [name for name in fields if name in dt.names]
-    formats = [dt.fields[name][0] for name in fields if name in dt.names]
-    offsets = [dt.fields[name][1] for name in fields if name in dt.names]
+    if ary.dtype.fields == None:
```
Don't mind the explicit check, could make it an assert if you want to allow for optimizing it away. It can't actually happen, since the indexing code only tries field access if the array has fields.
Frankly, I am not sure how we typically handle this.

ahaldane replied:
Hmm I could have sworn I added this to fix a bug. Maybe it was in an old version of this PR where _index_fields took care of more cases. In any case I'll remove it now.

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1783,35 +1780,35 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
-                PyArray_Descr *descr;
-                int offset;
-                PyObject *title;
+        if (!(PyString_Check(ind) || PyUnicode_Check(ind))) {
```
Is this Py3k save? It might be, we may have some of these macro'd (in which case some of the above can be simplified).
Could you possible add a test that tests for these type of unicode vs. string stuff (assuming they do not already exist)?

ahaldane replied:
Numpy does define a `PyUString_Check`, but that does not check for unicode in python2. In descriptor.c I see a lot of

    #if defined(NPY_PY3K)
            if (PyUString_Check(title)) {
    #else
            if (PyUString_Check(title) || PyUnicode_Check(title)) {
    #endif

I don't see the point of using `PyUString` there since in the top it always compiles to `PyUnicode_Check(title)` and in the bottom always to `PyString_Check(title)`.

I'll update this to just use the usual checks, but maybe in the future it might be nice to define a macro to simplify all those cases.

ahaldane replied:
Also, there are already lots of unit tests for Py3k and unicode, that's how I found out I needed to be more careful. These lines happened to already be safe because they are only reached after going through `obj_is_string_or_stringlist`, but it is clearer/better if I add checks too.

seberg replied:
Great, just noticed it used a different pattern and didn't think about it :).

seberg commented on numpy/core/src/multiarray/mapping.c:
```diff
@@ -1783,35 +1780,35 @@ array_assign_subscript(PyArrayObject *self, PyObject *ind, PyObject *op)
+        }
+
+        if (PyArray_CopyObject((PyArrayObject*)obj, op) < 0) {
+            Py_XDECREF(obj);
```
+1 for funky segfaults ;), this is a DECREF, not an XDECREF :)

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3575,20 +3575,19 @@ def collect_warning_types(f, *args, **kwargs):
         # All the different functions raise a warning, but not an error, and
         # 'a' is not modified:
-        assert_equal(collect_warning_types(a[['f1', 'f2']].__setitem__, 0, (10, 20)),
+        assert_equal(collect_warnings(a[['f1', 'f2']].__setitem__, 0, (10, 20)),
```
assert_warns is probably a bit nicer (no need for the local collect warnings), but we can put that off to another time.

ahaldane replied:
I'll leave it for another time. I was just fixing >80 char lines that were bugging me here.

seberg commented on numpy/core/tests/test_multiarray.py:
```diff
@@ -3606,11 +3605,18 @@ def test_record_no_hash(self):
         self.assertRaises(TypeError, hash, a[0])
 
+    def test_empty_structure_creation(self):
+        # make sure these do not raise errors (gh-5631)
```
Wow, but it makes sense, why not have no fields if you have a size.... Spaces after ":" though ;).

seberg commented:
Looks good. Some nitpicks and possibly tests for basic unicode access would be nice. But should be ready modulo that unicode thing (which is likely just be me confused). (or modulo some veto about the KeyError thing).

ahaldane commented:
I went ahead and moved all imports in `_internal.py` to the top. Otherwise fixed most little problems, and there are already lots of tests for unicode (which helped me catch unicode bugs here).

I'm not sure why a previous PR's commit is getting lumped in here.. maybe I messed up a rebase somehow?

Once the ValueError->KeyError change can go forward, I'll squash everything.

charris commented:
I think it would be better to save the change `ValueError->KeyError` for later. The second is preferable, but the first is OK and I suspect we are going to have enough campatibility problems with 1.10 without introducing one we are already aware of.

seberg commented:
@charris ok, @ahaldane sorry for suggesting to go ahead with it.  Could you separate the two things? Probably we can give it a shot and merge it into master when 1.10 is released/branched off. Also gives it a lot of time for complains to start before an actual release.

ahaldane commented:
all right, will do within the next day or two

charris commented:
If you rebase on master the extra PR will go away.

ahaldane commented:
Undid KeyError changes, rebased, squashed, and added a docstring to `_index_fields`.

charris commented:
@ahaldane Merged, thanks. If the behavior has changed, it would be good to document it, My understanding is that this mostly simplifies the code and fixes some bugs.

seberg commented:
Thanks all!
